### PR TITLE
Fbc Repository Datasource

### PR DESCRIPTION
**Pull request by nikitajoshi1**
# Description
Added Firmware Repository Datasource. User can filter based on the name of the Repository.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# ISSUE TYPE

- Feature Pull Request

##### RESOURCE OR DATASOURCE NAME
FBC_Repository Datasource

##### OUTPUT
output

##### ADDITIONAL INFORMATION
TestResult
Coverage
# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility

### CODE REVIEW

**Comment by nikitajoshi1** [comment]
**Note:** This PR was dependent on some data from https://github.com/dell/terraform-provider-ome/pull/98 so you will see two commits until PR https://github.com/dell/terraform-provider-ome/pull/98 is merged. Will rebase with main after that. Please review commit 8d2b8098feba6bfb60951827d3c35d8f36b73989 for this PR.

File: ome/datasource_firmware_baseline_compliance_repository_test.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"terraform-provider-ome/helper"
+	"terraform-provider-ome/utils"
+	"testing"
+
+	. "github.com/bytedance/mockey"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
```
Comment (doriac11) [inline ome/datasource_firmware_baseline_compliance_repository_test.go:25]: Please update to not use the SDK package: use this one instead "github.com/hashicorp/terraform-plugin-testing/helper/resource"

File: ome/datasource_firmware_catalog_test.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"terraform-provider-ome/helper"
+	"terraform-provider-ome/utils"
+	"testing"
+
+	. "github.com/bytedance/mockey"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
```
Comment (doriac11) [inline ome/datasource_firmware_catalog_test.go:25]: Same here use: "github.com/hashicorp/terraform-plugin-testing/helper/resource" (I am removing the sdk in my other PR

File: models/catalog.go [hunk 3]
```diff
@@ -152,6 +152,7 @@ type CatalogRepository struct {
 	RepositoryType        types.String `tfsdk:"repository_type"`
 	Source                types.String `tfsdk:"source"`
 	Username              types.String `tfsdk:"username"`
+	Password              types.String `tfsdk:"password"`
```
Comment (doriac11) [inline models/catalog.go:155]: I dont think we should include this in the model.

File: ome/datasource_firmware_baseline_compliance_repository.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	if errGet != nil {
+		resp.Diagnostics.AddError(
+			"Error Reading Repositories",
+			errGet.Error(),
+		)
+		return
+	}
+	var repos []models.RepositoryModel
+
+	if len(plan.Names) == 0 {
```
Comment (doriac11) [inline ome/datasource_firmware_baseline_compliance_repository.go:90]: Could some of this logic be moved to helper file?

File: examples/data-sources/ome_fbc_repository/datasource.tf [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+# Get details of all the firmware repositories 
+data "ome_fbc_repository" "fbc-repository-all" {
+
+}
+output "fbc-repository" {
+  value = data.ome_fbc_repository.fbc-repository-all
+}
+
+# data "ome_fbc_repository" "fbc-repository-name-filter" {
+#         names = [
```
Comment (doriac11) [inline examples/data-sources/ome_fbc_repository/datasource.tf:23]: Could you add a comment above the names filter to give some details about what happens if filter is empty or invalid.

**Review by doriac11 — APPROVED** [review]

**Review by doriac11 — DISMISSED** [review]

**Review by doriac11 — APPROVED** [review]

**Review by shekhar-j — DISMISSED** [review]

File: helper/fbc_repository_helper.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	for _, repo := range repos {
+		if contains(repoNames, repo.Name) {
+			filteredArray = append(filteredArray, repo)
+		}
+	}
+
+	return filteredArray
+}
+
+func contains(list []string, target string) bool {
```
[the remark this answers is not in the document]
Comment (rounak-adhikary) [reply helper/fbc_repository_helper.go]: Funny thing is: from go1.21, the standard lib also has this same method https://pkg.go.dev/slices#Contains

File: clients/fbc_repository.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+package clients
+
+import (
+	"fmt"
+	"terraform-provider-ome/models"
+)
+
+// GetAllRepositories retrieves all repositories from the given catalogs array.
+func (c *Client) GetAllRepositories(catalogsArray []models.CatalogsModel) (*[]models.RepositoryModel, error) {
```
Comment (rounak-adhikary) [inline clients/fbc_repository.go:22]: This function does not invoke any API calls. Is it possible to move this logic to the helper package?

Comment (rounak-adhikary) [reply clients/fbc_repository.go:22]: Also, is there any particular reason to use pointers to slices? Go slices are themselves basically pointers to some underlying array and can be passed by value without any significant overhead. So IMHO pointers to slices are best avoided whenever possible.

**Pull request merged**